### PR TITLE
Remove some 0.15 items from cirq.sim

### LINKED ISSUE #3898: Deprecation of abstract/protocol methods
We need a better way to deprecate abstract methods.

Context (#3860): 

@balopat  This ends up being a breaking change though for any simulator that was overriding `_simulator_iterator`. This is why I had left the call to `_simulator_iterator` there in the original PR #3650. Otherwise I'd have just inlined and deleted `_simulator_iterator` outright: it's a private method and it's unused elsewhere.

It doesn't sound like anyone has been broken by this, so maybe it's okay. (We can just delete `_simulator_iterator` in that case; it's not called by anyone). I think there's still an unmet need for cleanly deprecate abstract methods, or methods that were intended to be overridden. The base class has to continue calling them until the deadline or it will break child classes. I like the error-by-default, but maybe there's a way to silence these errors when retaining the call to a deprecated method is required, like here.

_Originally posted by @daxfohl in https://github.com/quantumlib/Cirq/pull/3860#discussion_r587578684_

### PR DESCRIPTION
**Pull request by daxfohl**
This covers most of the deprecations for `cirq.sim` in 0.15. The two remaining ones have some challenges:

* Removing `target_tensor` etc from the initializers and replacing with `initial_state`: The repr still emits `target_tensor` so it wouldn't round-trip.
* Removing `log_of_measurement_results`: Same thing. Plus we forgot to deprecate it in a couple places.

Also this closes #3898, as the pattern we used to deprecate `copy` without parameters works well.

### CODE REVIEW

File: cirq-core/cirq/sim/density_matrix_simulator.py [hunk 1]
```diff
@@ -294,6 +251,12 @@ class DensityMatrixStepResult(
             results, ordered by the qubits that the measurement operates on.
     """
 
+    @deprecated_parameter(
+        deadline='v0.16',
+        fix='Remove parameter `simulator` as it is no longer used.',
+        parameter_desc='simulator',
+        match=lambda args, kwargs: 'simulator' in kwargs or len(args) > 2,
+    )
```
Comment (MichaelBroughton) [inline cirq-core/cirq/sim/density_matrix_simulator.py:259]: Does this need a deprecation ?

**Review by MichaelBroughton — COMMENTED** [review]
This looks good to me, just one question:

File: cirq-core/cirq/sim/density_matrix_simulator.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/density_matrix_simulator.py:259]: It's a public class, so in principle we shouldn't change the constructor without a deprecation cycle. In reality there's no reason for anyone ever to create one of these outside the simulator though, so it's not likely anyone would be broken if we just deleted it.

**Comment by daxfohl** [comment]
@maffoo When I run the formatter locally it doesn't show any errors, but here it shows lots of errors. Is there something I need to do locally after pulling the latest?

**Comment by daxfohl** [comment]
@maffoo Okay I figured out how to update locally. Now the formatter check passes here, but I had to include a bunch of formatting changes in the PR.

**Comment by daxfohl** [comment]
@maffoo nvm, I see you have a separate PR for the formatting changes, so ultimately everything should line up, hopefully without any conflicts.

File: cirq-core/cirq/sim/density_matrix_simulator.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/density_matrix_simulator.py:259]: Did you have a preference? I can either delete this parameter now or leave it for next release. If the latter then this is ready to merge.

File: cirq-core/cirq/sim/simulator_base.py [hunk 2]
```diff
@@ -229,9 +214,7 @@ def _core_iterator(
                 except TypeError:
                     raise TypeError(f"{self.__class__.__name__} doesn't support {op!r}")
 
-            step_result = self._create_step_result(sim_state)
-            yield step_result
-            sim_state = step_result._sim_state
+            yield self._create_step_result(sim_state)
```
Comment (daxfohl) [inline cirq-core/cirq/sim/simulator_base.py:217]: @95-martin-orion note this also finally reverses the change that concerned you here: https://github.com/quantumlib/Cirq/pull/4110#pullrequestreview-702346996

File: cirq-core/cirq/sim/density_matrix_simulator.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/sim/density_matrix_simulator.py:259]: @MichaelBroughton is there anything holding this up now? I have a couple more deprecation PRs that depend on this one. I'd like this to go in soon if possible.

**Review by MichaelBroughton — APPROVED** [review]
LGTM

**Pull request merged**